core/xmp_parser.cpp

```diff
@@ -0,0 +1,137 @@
+#include "xmp_parser.h"
+#include "subsurface-string.h"
+#include "dive.h" // TODO: monster-include file only for utc_mktime.
+
+#include <libxml/parser.h>
+#include <libxml/tree.h>
+#include <cctype>
+
+static timestamp_t parse_xmp_date(const char *date)
+{
+	// Format: "yyyy-mm-dd[Thh:mm[:ss[.ms]][-05:00]]"
+	int year, month, day;
+	if (sscanf(date, "%d-%d-%d", &year, &month, &day) != 3)
+		return 0;
+
+	int hours = 0, minutes = 0, seconds = 0, milliseconds = 0;
+	int timezone = 0;
+
+	// Check for time part
+	if ((date = strchr(date, 'T')) != nullptr) {
+		++date; // Skip 'T'
+		if (sscanf(date, "%d:%d:%d.%d", &hours, &minutes, &seconds, &milliseconds) < 2)
+			return 0;
+
+		// Check for timezone part. Note that we simply ignore 'Z' as that
+		// means no time zone
+		while (*date && *date != '+' && *date != '-')
+			++date;
+		if (*date) {
+			int sign = *date == '+' ? 1 : -1;
+			int timezone_hours, timezone_minutes;
+			++date;
+			if (sscanf(date, "%d:%d", &timezone_hours, &timezone_minutes) != 2)
+				return 0;
+			timezone = sign * (timezone_hours * 60 + timezone_minutes) * 60;
+		}
+	}
+
+	// Round to seconds, since our timestamps are in seconds
+	if (milliseconds >= 500)
+		seconds += 1;
+
+	struct tm tm = { 0 };
+	tm.tm_year = year - 1900;
+	tm.tm_mon = month - 1;
+	tm.tm_mday = day;
+	tm.tm_hour = hours;
+	tm.tm_min = minutes;
+	tm.tm_sec = seconds;
+
+	timestamp_t res = utc_mktime(&tm);
+	res += timezone;
+
+	return res;
+}
+
+static timestamp_t extract_timestamp_from_attributes(const xmlNode *node)
+{
+	for (const xmlAttr *p = node->properties; p; p = p->next) {
+		const xmlChar *ns = p->ns ? p->ns->prefix : nullptr;
+
+		// Check for xmp::CreateDate property
+		if (!strcmp((const char *)ns, "xmp") && !strcmp((const char *)p->name, "CreateDate")) {
+			// We only support a single property value
+			if (!p->children || !p->children->content)
+				return 0;
+			const char *date = (const char *)p->children->content;
+			return parse_xmp_date(date);
+		}
+	}
+	return 0;
+}
+
+static timestamp_t extract_timestamp(const xmlNode *node)
+{
+	// We use a private stack, so that we can return in one go without
+	// having to unwind the call-stack. We only recurse to a fixed depth,
+	// since the data we are interested in are at a shallow depth.
+	// This can be increased on demand.
+	static const int max_recursion_depth = 16;
+	const xmlNode *stack[max_recursion_depth];
+	stack[0] = node;
+	int stack_depth = 1;
+
+	while (stack_depth > 0) {
+		const xmlNode *node = stack[stack_depth - 1];
+		// Parse attributes
+		timestamp_t timestamp = extract_timestamp_from_attributes(node);
+		if (timestamp)
+			return timestamp;
+
+		// Parse content, if not blank node. Content can only be at the second level,
+		// since it is always contained in a tag.
+		// TODO: We have to cast node to pointer to non-const, since we're supporting
+		// old libxml2 versions, where xmlIsBlankNode takes such a pointer. Remove
```

Comment:
how old? we build libxml2 on some platforms from source, anyway. But in order to understand how realistic it is to say "never mind", I'd need to know how old (or better, "how new") libxml2 has to be...

Reply:
I don't know - one of the Travis builds failed for this reason - but I don't remember which. I will push a private branch to find out.

Reply:
It's the linux and linux2 builds that fail. See https://travis-ci.org/Subsurface-divelog/subsurface/jobs/433133627
```
[ 48%] Building CXX object core/CMakeFiles/subsurface_corelib.dir/xmp_parser.cpp.o
/home/travis/build/Subsurface-divelog/subsurface/core/xmp_parser.cpp: In function ‘timestamp_t extract_timestamp(const xmlNode*)’:
/home/travis/build/Subsurface-divelog/subsurface/core/xmp_parser.cpp:95:27: error: invalid conversion from ‘const xmlNode* {aka const _xmlNode*}’ to ‘xmlNodePtr {aka _xmlNode*}’ [-fpermissive]
   if (!xmlIsBlankNode(node) && stack_depth >= 2) {
                           ^
In file included from /usr/include/libxml2/libxml/parser.h:16:0,
                 from /home/travis/build/Subsurface-divelog/subsurface/core/xmp_parser.cpp:4:
/usr/include/libxml2/libxml/tree.h:922:3: error:   initializing argument 1 of ‘int xmlIsBlankNode(xmlNodePtr)’ [-fpermissive]
   xmlIsBlankNode  (xmlNodePtr node);
   ^
make[2]: *** [core/CMakeFiles/subsurface_corelib.dir/xmp_parser.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```
Apparently failing version:
`Setting up libxml2-dev:amd64 (2.9.1+dfsg1-3ubuntu4.13)`

Windows works, which is:
`libxml-2.0, version 2.9.4`

I guess keeping the cast is just the simple thing to do...

Reply:
Well, but that means that as of 16.04 (the oldest we actually support), we should have the right version: https://packages.ubuntu.com/search?keywords=libxml2-dev

I'm fine with keeping the cast. It's annoying that Travis still doesn't offer an out of the box 16.04 or newer.